### PR TITLE
Forbid `server_info`-based labels in deny rules

### PR DESCRIPTION
**Pull request by atburke**
This change:
- Adds the `dynamic/` prefix to labels from a `server_info` resource create with `tctl`
- Forbids labels with the `dynamic/` prefix from being used in deny rules for new roles. Existing roles with `dynamic/` labels in deny rules, if any exist, will generate a cluster alert as well as a warning in `tctl`.

This is a breaking change, but creating `server_info` resources with `tctl` is not documented so we don't expect any users to be affected. See [this conversation](https://github.com/gravitational/teleport/pull/34475#discussion_r1411189547) for more details.

Changelog: Forbid labels prefixed with `dynamic/` from being used in deny rules

### CODE REVIEW

File: tool/tctl/common/resource_command.go
Comment (fspmarshall) [inline tool/tctl/common/resource_command.go:442]: nit: the error message returned by `CheckDynamicLabelsInDenyRules` is a bit vague.  Since this location is the most likely location where users will be forced to confront this error directly, having a more verbose message (like you're already doing with the alert) would probably be beneficial.

**Review by fspmarshall — APPROVED** [review]
Thanks for doing this!

File: lib/services/access.go
Comment (nklaassen) [reply lib/services/access.go:68]: That's quite tricky to do while parsing/before evaluating the expression, which is what we'd want. Label expressions use the [typical](https://github.com/gravitational/teleport/blob/047ecdfd60a165dbabb4aca3c020f27a0220c930/lib/services/label_expressions.go#L59) parser library which tries to be general-purpose and assumes everything including map keys could be subexpressions. So it would be very difficult to add some special case during the parse to check the allowed keys for this case. Especially while label expression support the [labels_matching](https://github.com/gravitational/teleport/blob/047ecdfd60a165dbabb4aca3c020f27a0220c930/lib/services/label_expressions.go#L115) function that can reference all labels where the key matches a regex

But, we could filter those labels out or return an error when the expression is actually being evaluated, pretty easy to add here https://github.com/gravitational/teleport/blob/047ecdfd60a165dbabb4aca3c020f27a0220c930/lib/services/label_expressions.go#L67. Unfortunately the user wouldn't see the error until they are actually trying to use the role.

You could try evaluating the expression with some dummy input to catch the error, but in general the evaluation could return an error for any number of reasons, even if the expression is perfectly valid, and I would prefer to avoid that

Comment (nklaassen) [reply lib/services/access.go:68]: There's also the option of doing a dumb regex search of the expression string for `"dynamic/`. I think it would be pretty unlikely to get false positives

File: lib/services/access.go
Comment (nklaassen) [inline lib/services/access.go:83]: Something like this to catch label expressions as well
```suggestion
	for _, kind := range []types.LabelMatcherKinds {
		labelMatchers, err := r.GetLabelMatchers(types.Deny, kind)
		for label := range labelMatchers.Labels {
			if strings.HasPrefix(label, types.TeleportDynamicLabelPrefix) {
				return trace.BadParameter(errorMessage)
			}
		}
		if strings.Contains(labelMatchers.Expression, `"` + types.TeleportDynamicLabelPrefix) {
				return trace.BadParameter(errorMessage)
		}
```

**Comment by atburke** [comment]
Friendly ping @nklaassen

File: lib/services/access_test.go
Comment (nklaassen) [inline lib/services/access_test.go:59]: Can you add a test case for the label expression check? Otherwise looks okay

**Review by rosstimothy — APPROVED** [review]
A few nits but LGTM otherwise. Please wait for @nklaassen to sign off as well before merging.

**Review by nklaassen — APPROVED** [review]

**Pull request merged**
